In [1]:
# ignore warnings
import warnings
warnings.filterwarnings("ignore")

# Data getting, cleaning, and exploring
import wrangle as w
import evaluate as ev
import explore as ex

# Python without these is hard
import pandas as pd
import numpy as np
from pydataset import data
from scipy import stats

# Machine Learning
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
import sklearn.preprocessing
from sklearn.preprocessing import MinMaxScaler 
from sklearn.impute import SimpleImputer
# Regression Modeling
from sklearn.metrics import mean_squared_error,r2_score
from sklearn.linear_model import LinearRegression, LassoLars, TweedieRegressor
from sklearn.preprocessing import PolynomialFeatures

# Visualization
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [2]:
# w.get_zillow_data(cached=False)

In [3]:
df = w.wrangle_zillow()

In [4]:
df.shape

(77413, 67)

In [5]:
df.parcelid.nunique()

77413

In [6]:
df = df[df.latitude.notnull()]

In [7]:
df = df[df.longitude.notnull()]

In [8]:
ex.missing_zero_values_table(df)

Your selected dataframe has 67 columns and 77380 Rows.
There are 52 columns that have NULL values.


Zero Values  NULL Values  \
buildingclasstypeid                     0        77365   
buildingclassdesc                       0        77365   
finishedsquarefeet13                    0        77339   
storydesc                               0        77330   
basementsqft                            0        77330   
storytypeid                             0        77330   
yardbuildingsqft26                      0        77310   
fireplaceflag                           0        77208   
architecturalstyletypeid                0        77174   
architecturalstyledesc                  0        77174   
typeconstructiontypeid                  0        77158   
typeconstructiondesc                    0        77158   
finishedsquarefeet6                     0        76994   
pooltypeid10                            0        76916   
decktypeid                              0        76766   
poolsizesum                             0        76513   
pooltypeid2                             0        76306   
hashottuborspa                          0        75842   
yardbuildingsqft17                      0        74992   
taxdelinquencyyear                      0        74494   
taxdelinquencyflag                      0        74494   
finishedsquarefeet15                    0        74371   
finishedfloor1squarefeet                0        71357   
finishedsquarefeet50                    0        71357   
fireplacecnt                            0        69104   
threequarterbathnbr                     0        67284   
pooltypeid7                             0        62338   
poolcnt                                 0        61243   
numberofstories                         0        59820   
airconditioningdesc                     0        52427   
airconditioningtypeid                   0        52427   
garagetotalsqft                      7685        51906   
garagecarcnt                          230        51906   
regionidneighborhood                    0        46490   
heatingorsystemtypeid                   0        27941   
heatingorsystemdesc                     0        27941   
buildingqualitytypeid                   0        27709   
propertyzoningdesc                      0        27050   
unitcnt                                 0        26818   
lotsizesquarefeet                       0         8239   
finishedsquarefeet12                    0         3632   
regionidcity                            0         1471   
calculatedbathnbr                       0          609   
fullbathcnt                             0          609   
yearbuilt                               0          263   
censustractandblock                     0          244   
calculatedfinishedsquarefeet            0          196   
structuretaxvaluedollarcnt              0          112   
regionidzip                             0           50   
taxamount                               0            5   
landtaxvaluedollarcnt                   0            2   
taxvaluedollarcnt                       0            1   
logerror                                0            0   
regionidcounty                          0            0   
fips                                    0            0   
rawcensustractandblock                  0            0   
parcelid                                0            0   
latitude                                0            0   
transactiondate                         0            0   
propertycountylandusecode               0            0   
id                                      0            0   
longitude                               0            0   
bedroomcnt                            830            0   
roomcnt                             59807            0   
assessmentyear                          0            0   
bathroomcnt                           592            0   
propertylandusetypeid                   0            0   

                              % of Total NULL Values  \
buildingclasstypeid          

In [9]:
# missing_cols_df = pd.Series(data=df.isnull().sum(axis = 1).value_counts().sort_index(ascending=False))

In [10]:
# missing_cols_df = pd.DataFrame(missing_cols_df)

In [11]:
# missing_cols_df = missing_cols_df.reset_index()

In [12]:
# missing_cols_df.columns = ['total_missing_cols','num_rows']

In [13]:
# missing_cols_df['percent_cols_missing'] = round(100 * missing_cols_df.total_missing_cols / df.shape[1], 2)

In [14]:
# missing_cols_df.head()

In [15]:
ex.missing_columns(df)

total_missing_cols  num_rows  percent_cols_missing  percent_rows_affected
0                   48         3                 71.64                   0.00
1                   47         3                 70.15                   0.00
2                   46         5                 68.66                   0.01
3                   45        50                 67.16                   0.06
4                   44        78                 65.67                   0.10
5                   43        28                 64.18                   0.04
6                   42        23                 62.69                   0.03
7                   41        29                 61.19                   0.04
8                   40       230                 59.70                   0.30
9                   39       285                 58.21                   0.37
10                  38      1925                 56.72                   2.49
11                  37      3620                 55.22                   4.68
12                  36      5775                 53.73                   7.46
13                  35      5148                 52.24                   6.65
14                  34     13326                 50.75                  17.22
15                  33     14782                 49.25                  19.10
16                  32     12579                 47.76                  16.26
17                  31      9891                 46.27                  12.78
18                  30      3455                 44.78                   4.46
19                  29      5270                 43.28                   6.81
20                  28       455                 41.79                   0.59
21                  27       316                 40.30                   0.41
22                  26        65                 38.81                   0.08
23                  25        24                 37.31                   0.03
24                  24        13                 35.82                   0.02
25                  23         2                 34.33                   0.00

In [16]:
df.propertylandusetypeid.value_counts().sort_index()

31.0        15
246.0     2009
247.0      535
248.0      727
260.0       37
261.0    52319
263.0       74
264.0        6
265.0      333
266.0    19294
267.0       29
269.0     1944
275.0       58
Name: propertylandusetypeid, dtype: int64

### propertylandusetypeid code meanings:
- <span style="color:red">red</span>: means that there are properties of this type in dataframe and are not "single unit properties"
- <span style="color:green">green</span>: are properties that are accepted as "single unit properties"
- ~~strikeout~~: are property types that are not included in the dataframe

    - <span style="color:red">31	Commercial/Office/Residential Mixed Used</span>
    - ~~46	Multi-Story Store~~
    - ~~47	Store/Office (Mixed Use)~~
    - <span style="color:red">246	Duplex (2 Units, Any Combination)</span>
    - <span style="color:red">247	Triplex (3 Units, Any Combination)</span>
    - <span style="color:red">248	Quadruplex (4 Units, Any Combination)</span>
    - <span style="color:green">260	Residential General</span>
    - <span style="color:green">261	Single Family Residential</span>
    - ~~<span style="color:green">262	Rural Residence</span>~~
    - <span style="color:green">263	Mobile Home</span>
    - <span style="color:green">264	Townhouse</span>
    - <span style="color:green">265	Cluster Home</span>
    - <span style="color:green">266	Condominium</span>
    - <span style="color:green">267	Cooperative</span>
    - ~~<span style="color:green">268	Row House</span>~~
    - <span style="color:red">269	Planned Unit Development</span>
    - ~~270	Residential Common Area~~
    - ~~271	Timeshare~~
    - ~~<span style="color:green">273	Bungalow</span>~~
    - ~~274	Zero Lot Line~~
    - <span style="color:green">275	Manufactured, Modular, Prefabricated Homes</span>
    - ~~<span style="color:green">276	Patio Home</span>~~
    - ~~<span style="color:green">279	Inferred Single Family Residential</span>~~
    - ~~290	Vacant Land - General~~
    - ~~291	Residential Vacant Land~~

In [17]:
df.propertylandusetypeid = df.propertylandusetypeid.astype(int)

In [18]:
# gave 'propertylandusetypeid' to variable x for ease, dropping propertylandusetypeid's that are not single unit properties
x='propertylandusetypeid'
df = df[df[x] != 31]
df = df[df[x] != 246]
df = df[df[x] != 247]
df = df[df[x] != 248]
df = df[df[x] != 269]

In [19]:
df.shape[0]

72150

In [20]:
df.propertylandusetypeid.value_counts().sort_index()

260       37
261    52319
263       74
264        6
265      333
266    19294
267       29
275       58
Name: propertylandusetypeid, dtype: int64

In [21]:
df.unitcnt.value_counts()

1.0    45325
2.0       33
3.0        3
4.0        2
6.0        1
Name: unitcnt, dtype: int64

In [22]:
df.unitcnt.isnull().sum()

26786

In [23]:
# fill unitcnt with 1 because these properties should be single unit properties

df.unitcnt.fillna(1, inplace=True)

In [24]:
df.unitcnt.isnull().sum()

0

In [25]:
df.unitcnt.value_counts()

1.0    72111
2.0       33
3.0        3
4.0        2
6.0        1
Name: unitcnt, dtype: int64

In [26]:
# drop any property listed less than or greater than 1 

df = df[df.unitcnt == 1]

In [27]:
df.unitcnt.value_counts()

1.0    72111
Name: unitcnt, dtype: int64

In [28]:
df.shape

(72111, 67)

In [29]:
df = w.handle_missing_values(df, 0.6, 0.75)
df.shape

(71961, 33)

In [30]:
df.isnull().sum().sort_values(ascending=False)

buildingqualitytypeid           27005
propertyzoningdesc              26597
heatingorsystemdesc             25386
heatingorsystemtypeid           25386
lotsizesquarefeet                8063
regionidcity                     1325
finishedsquarefeet12              255
censustractandblock               208
calculatedbathnbr                 119
fullbathcnt                       119
structuretaxvaluedollarcnt         84
yearbuilt                          72
regionidzip                        41
calculatedfinishedsquarefeet       38
taxamount                           5
taxvaluedollarcnt                   1
landtaxvaluedollarcnt               1
fips                                0
parcelid                            0
id                                  0
bathroomcnt                         0
bedroomcnt                          0
logerror                            0
longitude                           0
latitude                            0
assessmentyear                      0
propertyland

In [31]:
dropcols = ['id','heatingorsystemtypeid','buildingqualitytypeid','propertyzoningdesc','heatingorsystemdesc','calculatedbathnbr','regionidzip']

In [32]:
df = df.drop(columns=dropcols)

In [33]:
df[df.taxamount.isnull()]

parcelid  bathroomcnt  bedroomcnt  calculatedfinishedsquarefeet  \
18141  11487838          1.0         3.0                        1239.0   
21904  11647249          3.0         4.0                        2924.0   
22379  11666665          3.0         4.0                        2587.0   
46894  12920381          2.0         3.0                         944.0   
66936  14646889          2.0         2.0                        1030.0   

       finishedsquarefeet12    fips  fullbathcnt    latitude    longitude  \
18141                1239.0  6037.0          1.0  33874473.0 -118364035.0   
21904                2924.0  6037.0          3.0  34070466.0 -118462732.0   
22379                2587.0  6037.0          3.0  34048939.0 -118544920.0   
46894                 944.0  6037.0          2.0  34099255.0 -117891371.0   
66936                1030.0  6059.0          2.0  33613180.0 -117743363.0   

       lotsizesquarefeet  ... unitcnt  yearbuilt  structuretaxvaluedollarcnt  \
18141             6760.0  ...     1.0     1953.0                    188633.0   
21904                NaN  ...     1.0     1954.0                         NaN   
22379             9775.0  ...     1.0     1954.0                    448740.0   
46894             7489.0  ...     1.0     1930.0                     75000.0   
66936                NaN  ...     1.0     1974.0                     29646.0   

       taxvaluedollarcnt  assessmentyear  landtaxvaluedollarcnt  taxamount  \
18141           943166.0          2016.0               754533.0        NaN   
21904           461231.0          2016.0               461231.0        NaN   
22379          2030499.0          2016.0              1581759.0        NaN   
46894           352000.0          2016.0               277000.0        NaN   
66936            50450.0          2016.0                20804.0        NaN   

       censustractandblock  logerror  transactiondate  
18141         6.037621e+13  0.689382       2017-04-13  
21904         6.037265e+13  0.038192       2017-02-24  
22379         6.037263e+13 -0.043649       2017-01-25  
46894         6.037406e+13 -0.121397       2017-05-24  
66936         6.059063e+13 -0.013861       2017-03-14  

[5 rows x 26 columns]

In [34]:
df[df.taxvaluedollarcnt.isnull()]

parcelid  bathroomcnt  bedroomcnt  calculatedfinishedsquarefeet  \
53834  14006983          2.0         4.0                        1197.0   

       finishedsquarefeet12    fips  fullbathcnt    latitude    longitude  \
53834                1197.0  6059.0          2.0  33792921.0 -117982046.0   

       lotsizesquarefeet  ... unitcnt  yearbuilt  structuretaxvaluedollarcnt  \
53834             7210.0  ...     1.0     1954.0                         NaN   

       taxvaluedollarcnt  assessmentyear  landtaxvaluedollarcnt  taxamount  \
53834                NaN          2016.0                    NaN     388.18   

       censustractandblock  logerror  transactiondate  
53834         6.059088e+13  -0.03679       2017-07-31  

[1 rows x 26 columns]

In [35]:
df = df[df.taxvaluedollarcnt.notnull()]

In [36]:
df.isnull().sum().sort_values(ascending=False)

lotsizesquarefeet               8063
regionidcity                    1325
finishedsquarefeet12             255
censustractandblock              208
fullbathcnt                      119
structuretaxvaluedollarcnt        83
yearbuilt                         72
calculatedfinishedsquarefeet      38
taxamount                          5
transactiondate                    0
bathroomcnt                        0
bedroomcnt                         0
fips                               0
latitude                           0
longitude                          0
rawcensustractandblock             0
propertycountylandusecode          0
propertylandusetypeid              0
logerror                           0
regionidcounty                     0
roomcnt                            0
unitcnt                            0
taxvaluedollarcnt                  0
assessmentyear                     0
landtaxvaluedollarcnt              0
parcelid                           0
dtype: int64

In [37]:
df.structuretaxvaluedollarcnt.mean()

188021.7823643168

In [38]:
df.landtaxvaluedollarcnt.mean()

300497.03546414676

In [39]:
df.taxvaluedollarcnt.mean()

488301.9500138966

In [40]:
df.structuretaxvaluedollarcnt.mean() / df.landtaxvaluedollarcnt.mean()

0.6257026199073776

In [41]:
df[df.structuretaxvaluedollarcnt.isnull()]

parcelid  bathroomcnt  bedroomcnt  calculatedfinishedsquarefeet  \
3176    10832280          2.0         3.0                        1584.0   
3188    10832664          1.0         3.0                        1004.0   
3255    10835479          1.0         2.0                         869.0   
3256    10835480          1.0         2.0                        1036.0   
4356    10869709          3.0         4.0                        2990.0   
4363    10869828          4.0         3.0                        2598.0   
4366    10869942          3.0         3.0                        2175.0   
5536    10918175          3.0         4.0                        2692.0   
10675   11137835          4.0         3.0                        3284.0   
18220   11490443          2.0         4.0                        1416.0   
18451   11499034          1.0         2.0                         842.0   
18454   11499166          2.0         3.0                        1425.0   
18700   11508591          3.0         3.0                        2234.0   
19419   11540252          2.0         2.0                        1380.0   
19593   11548806          1.0         3.0                        1152.0   
19663   11551880          1.0         2.0                        1129.0   
19692   11553054          1.0         3.0                        1161.0   
19859   11562017          1.0         2.0                        1008.0   
20588   11595851          5.0         3.0                        4225.0   
20880   11608383          3.0         3.0                        2644.0   
20889   11608545          2.0         3.0                        2704.0   
20904   11609158          1.0         3.0                        1324.0   
21864   11645311          8.0         7.0                        9368.0   
21894   11646891          2.0         4.0                        1904.0   
21904   11647249          3.0         4.0                        2924.0   
21955   11649043          4.0         3.0                        2004.0   
22106   11656270          5.0         4.0                        4099.0   
22189   11658817          2.0         3.0                        1903.0   
22248   11660929          2.0         3.0                        2000.0   
22371   11666339          3.0         2.0                        2149.0   
22455   11668838          2.0         4.0                        1854.0   
22476   11669546          2.0         3.0                        1794.0   
22791   11687280          3.0         3.0                        2782.0   
24058   11755938          2.0         2.0                        1836.0   
26387   11905745          0.0         0.0                           NaN   
26484   11912966          1.0         2.0                        1360.0   
26641   11922646          1.0         2.0                         810.0   
26773   11928445          1.0         3.0                        1460.0   
26782   11928803          1.0         2.0                         850.0   
26802   11930466          1.0         2.0                         966.0   
28082   11997336          2.0         3.0                        1674.0   
28084   11997382          1.0         2.0                        1032.0   
28090   11997517          2.0         2.0                        1272.0   
28108   11998257          1.0         4.0                        1400.0   
28445   12014748          4.0         3.0                        1859.0   
28469   12015416          1.0         3.0                        1665.0   
29486   12055162          0.0         0.0                           NaN   
41762   12673310          1.0         3.0                        1145.0   
41857   12678523          1.0         1.0                         840.0   
43782   12765880          2.0         3.0                        1402.0   
48204   12987577          1.0         3.0                        1389.0   
48388   12999545          2.0         4.0                        1521.0   
48537   13009599          2.0    

### taxvaluedollarcnt = structuretaxvaluedollarcnt + landtaxvaluedollarcnt...  
- dropping values b/c
    - target feature if taxvaluedollarcnt and these values influence the target. 83 structure nulls
    - taxamount will be dropped as well. 5 taxamount nulls
    - so few values I am comfortable dropping

In [42]:
df = df[df.structuretaxvaluedollarcnt.notnull()]
df.shape

(71877, 26)

In [43]:
df = df[df.taxamount.notnull()]
df.shape

(71873, 26)

In [44]:
df.yearbuilt.describe()


count    71802.000000
mean      1969.341174
std         23.093796
min       1878.000000
25%       1954.000000
50%       1970.000000
75%       1987.000000
max       2016.000000
Name: yearbuilt, dtype: float64

In [45]:
# knn b/c close neighborhoods go up at similar dates
# will use yearbuilt, lat, long, fips
df.fips = df.fips.astype(int)
df.fips.dtype

dtype('int64')

In [48]:
df.isnull().sum()

parcelid                           0
bathroomcnt                        0
bedroomcnt                         0
calculatedfinishedsquarefeet      36
finishedsquarefeet12             243
fips                               0
fullbathcnt                      108
latitude                           0
longitude                          0
lotsizesquarefeet               8046
propertycountylandusecode          0
propertylandusetypeid              0
rawcensustractandblock             0
regionidcity                    1320
regionidcounty                     0
roomcnt                            0
unitcnt                            0
yearbuilt                         71
structuretaxvaluedollarcnt         0
taxvaluedollarcnt                  0
assessmentyear                     0
landtaxvaluedollarcnt              0
taxamount                          0
censustractandblock              194
logerror                           0
transactiondate                    0
dtype: int64

In [49]:
from sklearn.impute import KNNImputer

#Use numeric columns that you want to use for imputation
knn_cols_yr = df[['yearbuilt','latitude','longitude','fips']]

# define the thing
imputer = KNNImputer(n_neighbors=4)

# fit the thing (or fit and use with fit_transform) only on train!
imputed = imputer.fit_transform(knn_cols_yr)
imputed

array([[ 1.97200000e+03,  3.42225590e+07, -1.18617387e+08,
         6.03700000e+03],
       [ 1.97200000e+03,  3.42202610e+07, -1.18616409e+08,
         6.03700000e+03],
       [ 1.97200000e+03,  3.42224910e+07, -1.18616854e+08,
         6.03700000e+03],
       ...,
       [ 2.01000000e+03,  3.41050270e+07, -1.17805889e+08,
         6.03700000e+03],
       [ 1.97700000e+03,  3.39813000e+07, -1.17803849e+08,
         6.03700000e+03],
       [ 2.01400000e+03,  3.39150860e+07, -1.18372591e+08,
         6.03700000e+03]])

In [58]:
imputed = pd.DataFrame(imputed, index = df.index, dtype=int)
imputed.head()

0         1          2     3
0  1972  34222559 -118617387  6037
1  1972  34220261 -118616409  6037
2  1972  34222491 -118616854  6037
3  1960  34221864 -118615739  6037
4  1960  34220619 -118615253  6037

In [60]:
df['yearbuilt'] = imputed[[0]]

In [61]:
df.isnull().sum()

parcelid                           0
bathroomcnt                        0
bedroomcnt                         0
calculatedfinishedsquarefeet      36
finishedsquarefeet12             243
fips                               0
fullbathcnt                      108
latitude                           0
longitude                          0
lotsizesquarefeet               8046
propertycountylandusecode          0
propertylandusetypeid              0
rawcensustractandblock             0
regionidcity                    1320
regionidcounty                     0
roomcnt                            0
unitcnt                            0
yearbuilt                          0
structuretaxvaluedollarcnt         0
taxvaluedollarcnt                  0
assessmentyear                     0
landtaxvaluedollarcnt              0
taxamount                          0
censustractandblock              194
logerror                           0
transactiondate                    0
dtype: int64

In [63]:
df.censustractandblock[1]

60371132321007.0